internal/dcache/file_manager/file_manager.go

```diff
@@ -0,0 +1,361 @@
+		chunk, err := file.readChunkWithReadAhead(offset)
+		if err != nil {
+			return 0, nil
+		}
+		chunkOffset := getChunkOffsetFromFileOffset(offset, &file.FileMetadata.FileLayout)
+		copied := copy(buf[bufOffset:], chunk.Buf[chunkOffset:chunk.Len])
+		offset += int64(copied)
+		bufOffset += copied
+	}
+	common.Assert(offset <= file.FileMetadata.Size, "Read beyond the file size")
+	return bufOffset, nil
+}
+
+// Writes the file to the corresponing chunk(s) from the buf.
+func (file *DcacheFile) WriteFile(offset int64, buf []byte) error {
+	log.Debug("DistributedCache[FM]::WriteFile : offset : %d, bufSize : %d, file : %s", offset, len(buf), file.FileMetadata.Filename)
+
+	if offset < file.lastWriteOffset {
+		log.Info("DistributedCache[FM]::WriteFile: File handle is not writing in sequential manner, current Offset : %d, Prev Offset %d", offset, file.lastWriteOffset)
+		return syscall.ENOTSUP
+	}
+
+	endOffset := (offset + int64(len(buf)))
+	bufOffset := 0
+	for offset < endOffset {
+		chunk, err := file.CreateOrGetStagedChunk(offset)
+		if err != nil {
+			return err
+		}
+		chunkOffset := getChunkOffsetFromFileOffset(offset, &file.FileMetadata.FileLayout)
+
+		// Todo: Support Bigger Holes(hole size > chunkSize) inside the file if offset jumps suddenly to a higher offset.
+		// Sanity check for resetting the garbage data of the chunk buf.
+		if chunkOffset > chunk.Len {
+			resetBytes := copy(chunk.Buf[chunk.Len:chunkOffset], make([]byte, chunkOffset-chunk.Len))
+			chunk.Len += int64(resetBytes)
+		}
+
+		copied := copy(chunk.Buf[chunkOffset:], buf[bufOffset:])
+		offset += int64(copied)
+		bufOffset += copied
+		chunk.Len += int64(copied)
+
+		common.Assert(chunk.Len == getChunkOffsetFromFileOffset(offset, &file.FileMetadata.FileLayout),
+			fmt.Sprintf("Acutal Chunk Len : %d is modified incorrectly, Expected chunkLen : %d",
+				chunk.Len, getChunkStartOffsetFromFileOffset(offset, &file.FileMetadata.FileLayout)))
+
+		if chunk.Len == file.FileMetadata.FileLayout.ChunkSize {
+			// Schedule the upload
+			// todo: This not always true. if some writes to this chunk were skipped then there should be a
+			// way to stage this block.
+			scheduleUpload(chunk, file)
+		}
+	}
+	common.Assert(offset == endOffset, fmt.Sprintf("Write is not successful, expected Endoffset : %d, actual EndOffset : %d", endOffset, offset))
+	common.Assert(bufOffset == len(buf), fmt.Sprintf("Amount of Bytes copied : %d is not equal to buf len %d", bufOffset, len(buf)))
+	file.lastWriteOffset = offset
+	return nil
+}
+
+// Sync Buffers for the file with dcache/azure
+// This call can come when user application calls fsync()/close()
+func (file *DcacheFile) SyncFile() error {
+	log.Debug("DistributedCache[FM]::SyncFile : Sync File for %s", file.FileMetadata.Filename)
+	var err error
+	file.StagedChunks.Range(func(chunkIdx any, Ichunk any) bool {
+		chunk := Ichunk.(*StagedChunk)
+		scheduleUpload(chunk, file)
+		err = <-chunk.Err
```

Comment:
his will cause upload to be synchronized. We should try to do parallel uploads.
add a TODO for now, since sync is not a common case for us